wiring/inc/spark_wiring_cloud_event.h

```diff
@@ -0,0 +1,477 @@
+     *
+     * @return Event data.
+     */
+    Buffer data() const;
+
+    /**
+     * Get the event data.
+     *
+     * @return Event data.
+     */
+    String dataAsString() const;
+
+    /**
+     * Get the event data.
+     *
+     * @return Event data.
+     */
+    Variant dataAsVariant() /* FIXME: const */; // TODO: Rename?
+
+    /**
+     * Load the event data from a file.
+     *
+     * @param path File path.
+     * @return This event instance.
+     */
+    CloudEvent& loadData(const char* path);
+
+    /**
+     * Save the event data to a file.
+     *
+     * @param path File path.
+     * @return 0 on success, otherwise an error code defined by `Error::Type`.
+     */
+    int saveData(const char* path);
+
+    /**
+     * Check if the event has payload data.
+     *
+     * @return `true` if the event has payload data, otherwise `false`.
+     */
+    bool hasData() const {
+        return size() > 0;
+    }
+
+    /**
+     * Set the maximum size of event data that can be store on the heap.
+     *
+     * The data exceeding the specified size will be stored in a temporary file.
+     *
+     * This method has no effect if the event already contains any data.
+     *
+     * The default value is 1024 bytes.
+     *
+     * @param size Data size.
+     * @return This event instance.
+     */
+    CloudEvent& maxDataSizeInRam(size_t size);
+
+    /**
+     * Get the maximum size of event data that can be store on the heap.
+     *
+     * @return Data size.
+     */
+    size_t maxDataSizeInRam() const;
+
+    /**
+     * Set a callback to be invoked when the status of the event changes.
+     *
+     * @param callback Callback.
+     * @return This event instance.
+     */
+    CloudEvent& onStatusChange(std::function<OnStatusChange> callback);
+
+    /**
+     * Get the status of the event.
+     *
+     * @return Event status.
+     */
+    Status status() const;
+
+    /**
+     * Check if the event is valid.
+     *
+     * @return `true` if the event is valid, otherwise `false`.
+     */
+    bool valid() const {
+        return status() != Status::INVALID;
+    }
+
+    /**
+     * Check if the event is being sent to the Cloud.
+     *
+     * @return `true` if the event is being sent to the Cloud, otherwise `false`.
+     */
+    bool sending() {
+        return status() == Status::SENDING;
+    }
+
+    /**
+     * Check if the event was sent to the Cloud successfully.
+     *
+     * @return `true` if the event was sent successfully, otherwise `false`.
+     */
+    bool sent() const {
+        return status() == Status::SENT;
+    }
+
+    bool ok() const {
+        auto s = status();
+        return s != Status::FAILED && s != Status::INVALID;
+    }
+
+    int error() const;
+
+    /**
+     * Cancel sending the event.
+     *
+     * This method has no effect if the event is not currently being sent to the Cloud.
+     *
+     * A cancelled event cannot be published again.
```

Comment:
Is this still true? What state does a cancelled event go to?

Reply:
Yeah, a cancelled event gets invalidated. Making it transition to a failed state instead is possible and would be more user friendly but it will require some additional work: https://github.com/particle-iot/device-os/blob/7c898958ce7da201e0cc91c0b08180c35335aed8/wiring/src/spark_wiring_cloud_event.cpp#L415-L417.

Reply:
OK. No need for additional work here.